### ISSUE TITLE
2.0 Put file/get file issues

### ISSUE THREAD

**Issue by nitinjainsj** [pachyderm/pachyderm#6344]
```
$ pc version
COMPONENT           VERSION             
pachctl             2.0.0-alpha.21      
pachd               2.0.0-alpha.21      
$
```

A few put file issues...

- [x] Error message shows "test1.user@master" instead of "test1@master"
```
$ pc create repo test1
$ pc list file test1@master:
branches test1.user@master not found
```

- [x] Can not delete dir and all the files under the dir. Delete dir does not have equivalent "-r"
```
$ pc put file -r test1@master -f emptydir/
$ pc list file test1@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptyfile file 0B   
$ pc delete file test1@master:/emptydir
$ pc list file test1@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptyfile file 0B   
$ pc delete file test1@master:/emptydir/emptyfile
$ pc list file test1@master:/emptydir
NAME TYPE SIZE 
$
```

- [x] spurious dir. Seems like we are creating some sort of dir entry when the first command failed due to lack of "-f"
```
$ pc put file -r test@master: ./emptydir/
expected 1 arguments, got 2

Usage:
  pachctl put file <repo>@<branch-or-commit>[:<path/to/file>] [flags]

Examples:
....
$ pc put file -r test@master: -f ./emptydir/
emptydir/emptyfile 0.00 b / 0.00 b [---------------------------------------------] 0s 0.00 b/s
$ pc list file test@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptydir/ dir  0B   
$ pc list file test@master:/emptydir/emptydir
NAME                         TYPE SIZE 
/emptydir/emptydir/emptyfile file 0B   
$
$ ls -R emptydir/
emptyfile
$
```

- [x] Incorrect files created. After deleting the empty file, creating it again somehow remembered the last path which created /emptydir/emptydir
```
$ pc list file test@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptydir/ dir  0B   
$ pc delete file test@master:/emptydir/emptydir/emptyfile
$ pc list file test@master
NAME TYPE SIZE 
$ pc put file -r test@master: -f ./emptydir/
emptydir/emptyfile 0.00 b / 0.00 b [----------------------------------] 0s 0.00 b/s
$ pc list file test@master
NAME       TYPE SIZE 
/emptydir/ dir  0B   
$ pc list file test@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptydir/ dir  0B   
$ pc list file test@master:/emptydir/emptydir
NAME                         TYPE SIZE 
/emptydir/emptydir/emptyfile file 0B   
$ ls -R emptydir/
emptyfile
$
```

- [x] Since dir are not a thing, we need to give user feedback on the behavior that it is not a file
```
$ pc get file test@master:/emptydir
$
```

- [x] get file "-o" option for a dir cases stack trace (could not recreate it again)
```
$ pc get file test@master:/emptydir -o empyfile
empyfile 0.00 b / 0.00 b [-----------------------------] 0s 0.00 b/s
panic: increment required before ewma iteration update

goroutine 97 [running]:
github.com/vbauerster/mpb/v6.ewmaIterationUpdate(0x0, 0xc00075e480, 0x10cae)
	/Users/nitin/go/pkg/mod/github.com/vbauerster/mpb/v6@v6.0.2/bar.go:471 +0xda
github.com/vbauerster/mpb/v6.(*Bar).DecoratorEwmaUpdate.func1(0xc00075e480)
	/Users/nitin/go/pkg/mod/github.com/vbauerster/mpb/v6@v6.0.2/bar.go:236 +0x39
github.com/vbauerster/mpb/v6.(*Bar).serve(0xc000afe080, 0x310f0a8, 0xc000aee480, 0xc00075e480)
	/Users/nitin/go/pkg/mod/github.com/vbauerster/mpb/v6@v6.0.2/bar.go:297 +0x83
created by github.com/vbauerster/mpb/v6.newBar
	/Users/nitin/go/pkg/mod/github.com/vbauerster/mpb/v6@v6.0.2/bar.go:98 +0x368
$
```

- [x] weird behavior of pachctl get file for dir. 
```
$ pc list file test@master:/emptydir
NAME                TYPE SIZE 
/emptydir/emptydir/ dir  0B   
$ pc list file test@master:/emptydir/emptydir
NAME                         TYPE SIZE 
/emptydir/emptydir/emptyfile file 0B   
$ pc get file test@master:/emptydir/emptydir/emptyfile -o emptyfile
emptyfile 0.00 b / 0.00 b [-----------------------------] 0s 0.00 b/s
$ l
total 8
-rw-r--r--  1 nitin  staff  0 Jun 14 21:47 emptyfile
-rw-r--r--  1 nitin  staff  2 Jun 14 21:45 twobytefile
$ pc get file test@master:/emptydir/emptydir -o emptydir
emptydir 0.00 b / 0.00 b [-----------------------------] 0s 0.00 b/s
$ l
total 8
-rw-r--r--  1 nitin  staff  0 Jun 14 21:48 emptydir
-rw-r--r--  1 nitin  staff  0 Jun 14 21:47 emptyfile
-rw-r--r--  1 nitin  staff  2 Jun 14 21:45 twobytefile
$ pc get file test@master:/emptydir -o emptydir0
emptydir0 0.00 b / 0.00 b [-----------------------------] 0s 0.00 b/s
$ l
total 8
-rw-r--r--  1 nitin  staff  0 Jun 14 21:48 emptydir
-rw-r--r--  1 nitin  staff  0 Jun 14 21:48 emptydir0
-rw-r--r--  1 nitin  staff  0 Jun 14 21:47 emptyfile
-rw-r--r--  1 nitin  staff  2 Jun 14 21:45 twobytefile
$ 
```

[AB#23](https://dev.azure.com/PachydermCompany/92b98a32-88e5-46cb-a28f-5d064712e03b/_workitems/edit/23)

**Comment by brycemcanally** [comment]
1. The fully resolved repo, which includes the repo type (user), is shown when a branch is not found.
2. You disambiguate between a directory delete and file delete with a trailing slash. We are thinking about making this a more structured operation rather than relying on a string encoding that would be easy to forget or leave out. It probably makes sense to have a special field in the proto to indicate a directory deletion and to expose this through pachctl with the -r flag.
3. A directory is expected in this case because a file exists. We support empty files, not empty directories.
5. Everything looks correct here. You deleted the empty file which causes the two parent directories to not exist, then recreated it which causes the two parent directories to exist again.
6. This command is equivalent to getting all of the file content under `/emptydir`. The content is empty because there is only an empty file, so it returns nothing. This is expected behavior.
7. This seems like a bug in the progress library. It is probably related to it being a size zero file. This doesn't seem too high priority.
8. This is expected behavior, but may not be expected behavior for a user. @brendoncarroll and I have been discussing some changes to what get file does and I think we should extend the changes further to account for directories.

Most of this behavior is expected, but we can, and should, make some changes to make it easier for users. Most of this seems to revolve around the ergonomics for handling directories through pachctl, which I think we can address by adding corresponding -r flags where appropriate (delete, get). We want to address this at the gRPC layer by not depending on the string encoding to determine if we should operate on a directory and instead just have a field in the proto for indicating a directory operation.

**Comment by brendoncarroll** [comment]
Item 5 should be addressed by fixing #6324

**Comment by brendoncarroll** [comment]
The extra directory issue is fixed by https://github.com/pachyderm/pachyderm/pull/6350

**Comment by jdoliner** [comment]
I looked into the task "get file "-o" option for a dir cases stack trace (could not recreate it again)" and wasn't able to reproduce it. Furthermore I audited the code and couldn't find a way that we could trigger that issue. I'm still a little wary because obviously it was triggered somehow but I can't see any way to get it to trigger. If someone does hit the issue they can always use `--no-progress` to get around it and if it's reproducible feel free to post the repro here and we can fix it.
